xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
-			ClusterType:             xdsresource.ClusterTypeAggregate,
-			ClusterName:             "cluster" + fmt.Sprint(i),
-			PrioritizedClusterNames: []string{"cluster" + fmt.Sprint(i+1)},
-		}, nil)
-		if i == 15 {
-			// The 16th iteration will try and create a cluster which exceeds
-			// max stack depth and will thus error, so no CDS Watch will be
-			// started for the child.
-			continue
-		}
-		_, err = fakeClient.WaitForWatchCluster(ctx)
-		if err != nil {
-			t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("Timed out waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
 		}
 	}
-	select {
-	case chu := <-ch.updateChannel:
-		if chu.err.Error() != "aggregate cluster graph exceeds max depth" {
-			t.Fatalf("Did not receive the expected error, instead received: %v", chu.err.Error())
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for an error to be written to update channel.")
-	}
-}
 
-// TestDiamondDependency tests a diamond shaped aggregate cluster (A->[B,C];
-// B->D; C->D). Due to both B and C pointing to D as it's child, it should be
-// ignored for C. Once all 4 clusters have received a CDS update, an update
-// should be then written to the update buffer, specifying a single Cluster D.
-func (s) TestDiamondDependency(t *testing.T) {
-	ch, fakeClient := setupTests()
-	ch.updateRootCluster("clusterA")
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	_, err := fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
-	}
-	fakeClient.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{
-		ClusterType:             xdsresource.ClusterTypeAggregate,
-		ClusterName:             "clusterA",
-		PrioritizedClusterNames: []string{"clusterB", "clusterC"},
-	}, nil)
-	// Two watches should be started for both child clusters.
-	_, err = fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
-	}
-	_, err = fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	const wantErr = "aggregate cluster graph exceeds max depth"
+	client := testgrpc.NewTestServiceClient(cc)
+	_, err := client.EmptyCall(ctx, &testpb.Empty{})
+	if code := status.Code(err); code != codes.Unavailable {
+		t.Fatalf("EmptyCall() failed with code: %v, want %v", code, codes.Unavailable)
+	}
+	if err != nil && !strings.Contains(err.Error(), wantErr) {
+		t.Fatalf("EmptyCall() failed with err: %v, want err containing: %v", err, wantErr)
+	}
+
+	// Start a test service backend.
+	server := stubserver.StartTestService(t, nil)
+	t.Cleanup(server.Stop)
+
+	// Update the aggregate cluster resource to no longer exceed max depth.
```

Comment:
Mention it's right on the boundary. And thus should be valid and root cluster eventually used for RPC.